click/_winconsole.py

```diff
@@ -209,13 +233,21 @@ def _get_text_stdin(buffer_stream):
 
 
 def _get_text_stdout(buffer_stream):
+    if PY2:
+        buffer_stream = StreamWrapper(buffer_stream)
+        sys.stdout = StreamWrapper(sys.stdout)
```

Comment:
`buffer_stream` will always be `sys.stdout`, so we should be able to assign the same wrapper to both, right?

Reply:
Wrapping `sys.{stdout,stderr}` should probably be moved elsewhere actually... I think this can be called multiple times? Wrapping `sys.{stdout,stderr}` should be done only once. But I'm not sure where. It starts to become necessary to do so once you call `setmode(..., O_BINARY)`. And if this can be called multiple times it's possible `buffer_stream` will already be a `StreamWrapper`...

This needs some place that will only be called once or otherwise we should remember if we already applied the patch.

Reply:
I think multiple calls are either cached or exit early, but I'd have to check again.

Maybe just extract the `O_BINARY` part to a common function that checks `isinstance(stream, StreamWrapper)`.

Reply:
`isinstance(stream, StreamWrapper)` will probably interact badly with other libraries that touch `sys.std*`. Calling `get_text_stream` from the utils module probably runs this again.

There are two places we call `setmode`: In `_get_windows_console_stream` here, and in `set_binary_mode` in _compat.py. But those only get the file object and won't know which global to patch. They are only called by functions in _compat.py which are per std stream. I can add a function, and call it from these places, which get's the name of the stream and if it is the right Windows version, Python version and we didn''t patch that stream already (Module level variable), then it will apply the patch.